File: TinyScript/java/src/main/java/paser/ast/AssignStmt.java
package paser.ast;

import lexer.Token;
import paser.util.ParseException;
import paser.util.PeekTokenIterator;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/8/23 0:21
 **/
public class AssignStmt extends Stmt {
    protected AssignStmt() {
        super(ASTNodeTypes.ASSIGN_STMT, "assign");
    }

    public static ASTNode parse(ASTNode parent, PeekTokenIterator it) throws ParseException {
        AssignStmt assignStmt = new AssignStmt();
        Token tkn = it.peek();
        ASTNode factor = Factor.parse(parent, it);
        if (factor == null || tkn.isScalar()) {
            throw new ParseException(tkn);
        }
        assignStmt.addChild(factor);
        Token token = it.nextMatch("=");
        ASTNode parse = Expr.parse(it);
        assignStmt.addChild(parse);
        assignStmt.setLexeme(token);
        return assignStmt;
    }
}

File: TinyScript/js/src/main/js/parser/ast/AssignStmt.ts
import Stmt from "./Stmt";
import ASTNodeTypes from "./ASTNodeTypes";
import ParseException from "../ParseException";
import PeekTokenIterator from "../PeekTokenIterator";

export default class AssignStmt extends Stmt {

    constructor() {
        super(ASTNodeTypes.ASSIGN_STMT, "assign");
    }

    static parse(it: PeekTokenIterator):AssignStmt {
        const {Factor, Expr} = require('./index');
        const assign = new AssignStmt();
        const tkn = it.peek();
        const factor = Factor.parse(it);
        if (factor == null && tkn != null) {
            throw ParseException.fromToken(tkn);
        }
        assign.addChild(factor);
        const lexeme = it.nextMatch('=');
        const expr = Expr.parse(it);
        assign.addChild(expr);
        assign.lexeme = lexeme;
        return assign;
    }
}